### PR TITLE
Simple change for future multi-system to Navier Stokes advected quantities

### PR DESCRIPTION
**Pull request by GiudGiud**
I gutted this PR to just leave the cleaning up of derivatives from RC UO. 
Is this something we want to handle there?

Or should it be done elsewhere?
- functors could check the current system being solved before returning AD numbers with derivatives
- another idea

OUTDATED:

A few things to discuss:
- should we allow to separate all momentum components in different systems? It seems a common thing to do, but we dont have the Picard fixed point executioner to make it work rn @grmnptr 
- should we separate scalars by default? For reactor physics it make sense. But maybe species react in other physics?

Some preliminary results show this is a lot faster than fully coupled (no shocker here) and possibly 20% faster than multiapps with maybe 60% of the memory cost (no duplication of mesh, a_xy, P, velocities). 

This is still missing some work to enable:
- [ ] transients
- [ ] scaling (see https://github.com/idaholab/moose/issues/25646)
- [ ] cleaner RC code (and maybe faster, ditching AD on the non-velocity system)
- [ ] passing tests

END OUTDATED

### CODE REVIEW

**Comment by grmnptr** [comment]
The momentum-pressure coupling is special, we cannot seperate them just like that. That is coming in the SIMPLE PR which will have the option to separate momentum components. It would be cool if we could merge these efforts into a single executioner or two executors which interact. I'd definitely segregate every scalat but mauricio might have cases where a monolithic could be better.

**Comment by GiudGiud** [comment]
Sounds good.
I can force mass & momentum in the same system here through the parameters. I was planning to focus my efforts on Transients from now on.

**Comment by GiudGiud** [comment]
overall this can be buried but there s a few things I could port

**Comment by GiudGiud** [comment]
Updated the description after I gutted the PR

I think in the near future we can address https://github.com/idaholab/moose/issues/24691
for fully coupled NS + precursor advection

To make it in NS we need some code in the RC UO to remove the derivatives.

File: modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C [hunk 1]
```diff
@@ -540,7 +551,12 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
         &fi, Moose::FV::LimiterType::CentralDifference, true, correct_skewness, boundary_elem};
     auto velocity = vel(boundary_face, time);
     incorporate_mesh_velocity(boundary_face, velocity);
-    return velocity;
+
+    // If not solving for velocity, clear derivatives
+    if (_fe_problem.currentNlSysNum() != _u->sys().number())
+      return {velocity(0).value(), velocity(1).value(), velocity(2).value()};
```
Comment (grmnptr) [inline modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:557]: I usually just use `raw_value` I think that would make it a little nicer. Same for the other changes.

**Review by grmnptr — CHANGES_REQUESTED** [review]
Great optimization idea!

File: modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C (hunk 1 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:557]: yeah I have no preference

**Comment by GiudGiud** [comment]
It crashes pretty nastily if you dont do this btw when trying multi-system in NS btw, it s not just optimization

File: modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C (hunk 1 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C]: changed

**Review by grmnptr — APPROVED** [review]
Nice! I believe the CI failures are not related.

File: modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C [hunk 2]
```diff
@@ -351,8 +351,15 @@ INSFVRhieChowInterpolator::initialize()
...
-    _a[pair.first] = 0;
+  // Dont reset if not in current system
+  // IDEA: clear them derivatives
+  if (_u->sys().number() == _fe_problem.currentNlSysNum())
+    for (const auto & pair : _a)
+      _a[pair.first] = 0;
+  else
+    for (const auto & pair : _a)
+      _a[pair.first] = {
+          _a[pair.first](0).value(), _a[pair.first](1).value(), _a[pair.first](2).value()};
```
Comment (lindsayad) [inline modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: The guy who made this code was really dumb (it was me)... I am iterating over the key-value pairs and then I do a lookup with the key when I already have access to the value?! This code can definitely be improved in multiple ways. I will put up a PR

Comment (GiudGiud) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: interesting that we 0 in one case and set the raw value in another. 
did I do that part?

Comment (lindsayad) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: lol in the PR you made that got merged two hours ago? Yes in this PR you added the raw value part (without using `raw_value` which #25898 will switch to).

Comment (GiudGiud) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: Yeah. Seems that PR was wrong

Comment (lindsayad) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: I'm confused. You're saying that the *this* PR is wrong?

Comment (GiudGiud) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: Yeah I think so. 
Probably bad copy paste I ll check when I have a machine ready

Comment (GiudGiud) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:362]: nevermind it s correct

**Comment by MengnanLi91** [comment]
@GiudGiud I saw you tagged my PR here. Could you provide more details on what you want for scaling? I only looked into the scaling factor in adjoint system.

**Comment by GiudGiud** [comment]
I was hitting issues with Steady2 (the 2 system test executioner) + scaling.

**Comment by MengnanLi91** [comment]
> I was hitting issues with Steady2 (the 2 system test object executioner) + scaling.

Do you have a test case I can try with my change in scaling PR?

**Comment by GiudGiud** [comment]
not with me but you can add scaling to this tests/misc/multiple-nl-systems/test-fv.i input.i
if it works with both manual and automatic scaling we are good to go

**Comment by MengnanLi91** [comment]
> not with me but you can add scaling to this tests/misc/multiple-nl-systems/test-fv.i input.i if it works with both manual and automatic scaling we are good to go

Yep, that error message is fixed! Both manual and automatic scaling work.

**Comment by GiudGiud** [comment]
Awesome! This is a great step for multi-system.

**Comment by GiudGiud** [comment]
def newsletter worthy ;)

**Pull request merged**